### ISSUE TITLE
kotlin_rules and Ijars: Update the Ijar tooling to do the right thing.

### ISSUE THREAD

**Issue by hsyed** [bazelbuild/bazel#4549]
I have been working on a new set of kotlin rules and they are currently being migrated to bazelbuild/rules_kotlin.

One goal with the design of the Kotlin JVM ruleset is to stay as close to the native java rules and infrastructure as possible. The automatic Ijars generation is  problematic as the java rules auto create ijars. Ijars generated for Kotlin work in a lot of cases, but inlined methods break (might be other cases). 

## Proposal: Adapt Ijar to do the right thing for Kotlin Jars.

I propose making a change to the behaviour of the ijar tool so that it does the right thing when it comes to Kotlin metadata. Adding this change to Bazel core simplifies the Kotlin rules and allows more reuse out of java common infrastructure. I propose the ijar tool be updated to:

1) Initially to detect a Kotlin jar and ignore it for processing -- perhaps just symlink the ijar to the full jar ? This should be straight forward a Kotlin jar has some markers that can be identified without processing bytecode.
2) Adapt the tool to generate valid Kotlin Ijars. Afaik this means retaining certain bytecode elements and marker files.

## Motivating example 1: dep management tooling / `java_import`.

I have [just added a `kotlin_import`](https://github.com/hsyed/rules_kotlin/blob/b1c6cbeb2d8943100745fad44d566d288d582fcb/kotlin/kotlin.bzl#L293-L326) rule to:

1) Get around the automatic ijar generation by `java_import`, and
2) Ensure the Jar has a `KotlinInfo` provider so that `full_compile_jars` are always selected when compiling Kotlin. 

Any fancy java dependency management tooling that we develop will likely work via orchestrations of `java_import`, `java_library` and `java_import_external` it's a shame to have to add a layer for Kotlin when it's barely needed.

**Comment by meteorcloudy** [comment]
@iirina Can you take a look at this?

**Comment by iirina** [comment]
Can you explain why the automatic ijar generation in the java rules is problematic? Is it because you are using `java_import` to import kotlin jars, in which case ijar will fail (sometimes)?

**Comment by hsyed** [comment]
I misrepresented the sometimes part. Most Kotlin libraries that a JVM project would use as an external dependency will have DSLs that make heavy use of `inline`. So some intermediate `kotlin_` rule has to be used to:

* Add a KotlinInfo -- to enable rules in the workspace to ensure `full_compile_jars` are used when using the target as a dependency.
* Force selection of `full_compile_jars` by re-interpreting the `JavaInfo` deps that are Kotlin Jars --e.g., using the `//jar` component of `maven_jar`. 

If the Kotlin ABI elements were retained by the Ijar tool transparently then special handling would not be needed so Kotlin libraries could be imported by  `java_import`, `maven_jar`,  `java_library`.

Current dependency management strategies (pubref/rules_maven, bazel-dep, migration-tooling, [this one](https://github.com/bazelbuild/bazel/pull/3946)) go beyond this. They generate BUILD file entries which range from from simply listing `runtime_deps` to modelling the transitive dep in fine grained detail (exports etc).

Current and future versions of these tools could work for Kotlin transparently as long as the ijar tool doesn't strip off the Kotlin ABI elements.

This goes beyond just the motivational example I gave above. The intellij plugin model has to declare and have conditional checks for intermediate rules such as `kotlin_import` and `kotlin_stdlib`. Unnecessary skylark logic has to be added to the rules.

**Comment by cushon** [comment]
> Adapt the tool to generate valid Kotlin Ijars. Afaik this means retaining certain bytecode elements and marker files.

Is there a reference for specifically how these ABI elements are represented in bytecode?

**Comment by hsyed** [comment]
Just had a closer look.

For marker files there is a [`META-INF/<module name>.kotlin_module`](https://blog.jetbrains.com/kotlin/2015/06/improving-java-interop-top-level-functions-and-properties/). Might be more depending on the compilation modes / module types.

After some experimentation inlining metadata seems to be embedded in the [Metadata](https://github.com/JetBrains/kotlin/blob/master/core/runtime.jvm/src/kotlin/Metadata.kt) annotation. Every class has a Metadata annotation. 

Using the bytecode tool included in the Intellij Kotlin plugin to inspect the following test file:
```kt
package test

fun monkey(x: () -> Unit) {
    x()
}

inline fun donkey(y: () -> Unit): Int {
    y()
    return 1
}

fun lonkey(z: () -> Unit): Int {
    z()
}
```

d2 entry in metadata -- notice the inlined function includes the param type. 
```
d2={"donkey", "", "y", "Lkotlin/Function0;", "", "lonkey", "z", "monkey", "x", "production sources for module .workspace"})
```

full output
```
// ================test/TestKt.class =================
// class version 50.0 (50)
// access flags 0x31
public final class test/TestKt {


  // access flags 0x19
  // signature (Lkotlin/jvm/functions/Function0<Lkotlin/Unit;>;)V
  // declaration: void monkey(kotlin.jvm.functions.Function0<kotlin.Unit>)
  public final static monkey(Lkotlin/jvm/functions/Function0;)V
    @Lorg/jetbrains/annotations/NotNull;() // invisible, parameter 0
   L0
    ALOAD 0
    LDC "x"
    INVOKESTATIC kotlin/jvm/internal/Intrinsics.checkParameterIsNotNull (Ljava/lang/Object;Ljava/lang/String;)V
   L1
    LINENUMBER 4 L1
    ALOAD 0
    INVOKEINTERFACE kotlin/jvm/functions/Function0.invoke ()Ljava/lang/Object;
    POP
   L2
    LINENUMBER 5 L2
    RETURN
   L3
    LOCALVARIABLE x Lkotlin/jvm/functions/Function0; L0 L3 0
    MAXSTACK = 2
    MAXLOCALS = 1

  // access flags 0x19
  // signature (Lkotlin/jvm/functions/Function0<Lkotlin/Unit;>;)I
  // declaration: int donkey(kotlin.jvm.functions.Function0<kotlin.Unit>)
  public final static donkey(Lkotlin/jvm/functions/Function0;)I
    @Lorg/jetbrains/annotations/NotNull;() // invisible, parameter 0
   L0
    ALOAD 0
    LDC "y"
    INVOKESTATIC kotlin/jvm/internal/Intrinsics.checkParameterIsNotNull (Ljava/lang/Object;Ljava/lang/String;)V
   L1
    LINENUMBER 8 L1
    ALOAD 0
    INVOKEINTERFACE kotlin/jvm/functions/Function0.invoke ()Ljava/lang/Object;
    POP
   L2
    LINENUMBER 9 L2
    ICONST_1
    IRETURN
   L3
    LOCALVARIABLE y Lkotlin/jvm/functions/Function0; L0 L3 0
    LOCALVARIABLE $i$f$donkey I L0 L3 1
    MAXSTACK = 2
    MAXLOCALS = 2

  // access flags 0x19
  // signature (Lkotlin/jvm/functions/Function0<Lkotlin/Unit;>;)I
  // declaration: int lonkey(kotlin.jvm.functions.Function0<kotlin.Unit>)
  public final static lonkey(Lkotlin/jvm/functions/Function0;)I
    @Lorg/jetbrains/annotations/NotNull;() // invisible, parameter 0
   L0
    ALOAD 0
    LDC "z"
    INVOKESTATIC kotlin/jvm/internal/Intrinsics.checkParameterIsNotNull (Ljava/lang/Object;Ljava/lang/String;)V
   L1
    LINENUMBER 13 L1
    ALOAD 0
    INVOKEINTERFACE kotlin/jvm/functions/Function0.invoke ()Ljava/lang/Object;
    POP
   L2
    LINENUMBER 14 L2
    ICONST_0
    IRETURN
   L3
    LOCALVARIABLE z Lkotlin/jvm/functions/Function0; L0 L3 0
    MAXSTACK = 2
    MAXLOCALS = 1

  @Lkotlin/Metadata;(mv={1, 1, 9}, bv={1, 0, 2}, k=2, d1={"\u0000\u0014\n\u0000\n\u0002\u0010\u0008\n\u0000\n\u0002\u0018\u0002\n\u0002\u0010\u0002\n\u0002\u0008\u0005\u001a\u0017\u0010\u0000\u001a\u00020\u00012\u000c\u0010\u0002\u001a\u0008\u0012\u0004\u0012\u00020\u00040\u0003H\u0086\u0008\u001a\u0014\u0010\u0005\u001a\u00020\u00012\u000c\u0010\u0006\u001a\u0008\u0012\u0004\u0012\u00020\u00040\u0003\u001a\u0014\u0010\u0007\u001a\u00020\u00042\u000c\u0010\u0008\u001a\u0008\u0012\u0004\u0012\u00020\u00040\u0003\u00a8\u0006\u0009"}, d2={"donkey", "", "y", "Lkotlin/Function0;", "", "lonkey", "z", "monkey", "x", "production sources for module .workspace"})
  // compiled from: test.kt
}


// ================META-INF/production sources for module .workspace.kotlin_module =================
	

testTestKt
```

I think I read somewhere that the bytecode contained within the stub of an inlined function is used as well.

references:

[understand every line of your codebase](https://resources.jetbrains.com/storage/products/kotlinconf2017/slides/Kotlin+Conf.pdf)

[kotlin-bytecode-generation-and-runtime-performance](https://www.slideshare.net/intelliyole/kotlin-bytecode-generation-and-runtime-performance)

**Comment by hsyed** [comment]
Yes I think thats right, the compiler seems to generate bytecode mimicking a regular JDK8 higher order function, the lambda invocation instructions must be used as the placeholders, nice design. 

It seems like a simple enough change, visit classes with Metadata annotations, discover public and protected inline methods and preclude them from bytecode dropping. 

I could gather a list of the file patterns relevant to the various module types -- as globs ?.

---

Since we are visiting the jars and the metadata I wonder if it would be possible to either:

1) Attach providers marking the Kotlin Jar type, along with some basic fields such as module name.
2) Or just attach a basic provider marking the Kotlin Jar type and write out a bazel proto file containing relevant metadata. 

There are going to be quite a few rule and rule type variants. If we had option 1 it would really simplify the design as we can constraint the attributes.

**Comment by lberki** [comment]
In my uninformed opinion, I think it's fine to special-case Kotlin stuff in ijar a little bit; the above "don't drop bytecode from methods that are marked as inline" counts as such in my book.

I don't understand the latter half of @hsyed 's comment, though -- how is this new proposed provider relevant to changes to ijar?

**Comment by hsyed** [comment]
@iberki I've removed the second part, it's not relevant to this issue at all.

**Comment by davidstanke** [comment]
Great -- sounds like we're well aligned on approach. @hsyed do you have the bandwidth to implement this? @cushon any concerns about the approach described?

**Comment by lberki** [comment]
@hsyed  -- excellent, I was afraid for a second that my understanding is *that* hazy :)

**Comment by hsyed** [comment]
@davidstanke I haven't looked at C++ in a long time and I just opened the JVM bytecode spec last week. I think it would be best implemented by someone who has already worked on the tool. 

There are two parts to this feature:

1) Retaining metadata files.
2) Discovery and retention of method bodies of inline functions.

1 is quite straight forward and I could implement this, but if someone else does 2 might as well get 1 one done in that PR.

**Comment by hsyed** [comment]
How would we test the changes ? I can add a test case to rules_kotlin that tests if the inlined methods are retained via a `java_import`.

**Comment by lberki** [comment]
That's one option. You could also add a pre-compiled Kotlin .jar somewhere under

`https://github.com/bazelbuild/bazel/tree/master/third_party/ijar/test`

and add a test that checks what happens to the class files. Or both. Both is probably the best.

**Comment by hsyed** [comment]
I've added some test cases, these cover the inlining change and retention of the metadata. 

Now onto the test. I can add a function to `ijar_test.sh` this could: 

1) Run the `ijar` tool on the jar. 
2) Use the` jar` tool to verify the metadata file has been retained.
3) Try to compile a kotlin client file that references the inline functions.

@lberki why do you think ?

**Comment by lberki** [comment]
I'd rather live without (3) because that would make the tests of ijar dependent on Kotlin. Other than that, awesome.

**Comment by hsyed** [comment]
rules_kotlin will be a part of the CI runs soon. It should catch kotlinc's opinion on the processed ijar.

For 3 we could assert function bodies still exist using `javap -c` ? 

```
Compiled from "InlineCases.kt"
public final class ClassWithInline {
  public final void classInlineFun(kotlin.jvm.functions.Function0<kotlin.Unit>);
    Code:
       0: aload_1
       1: ldc           #9                  // String op
       3: invokestatic  #15                 // Method kotlin/jvm/internal/Intrinsics.checkParameterIsNotNull:(Ljava/lang/Object;Ljava/lang/String;)V
       6: aload_1
       7: invokeinterface #21,  1           // InterfaceMethod kotlin/jvm/functions/Function0.invoke:()Ljava/lang/Object;
      12: pop
      13: return

  public ClassWithInline();
    Code:
       0: aload_0
       1: invokespecial #30                 // Method java/lang/Object."<init>":()V
       4: return
}
```

asserting this fragment should be enough of a test:

```
  public final void classInlineFun(kotlin.jvm.functions.Function0<kotlin.Unit>);
    Code:
```

**Comment by lberki** [comment]
Yep, checking the output of `javap -c` works; what I'd like to avoid is that the tests of ijar have a dependency on Kotlin. If you think that test is necessary, you can put it in `rules_kotlin`.

**Comment by cushon** [comment]
The information in metadata about which methods are inlined appears to be a serialized proto: https://github.com/JetBrains/kotlin/blob/ee50aec7342ab375a105bb318e6501135e859028/core/deserialization/src/descriptors.proto. @lberki are we OK with added a dependency on proto to ijar? I think there were some concerns about making the c++ singlejar depend on proto, but I'm not sure where that landed.

**Comment by hsyed** [comment]
I've wired up the tests in the pr (#4632), the ijar step is currently just cp'ing so it can be merged in.

**Comment by lberki** [comment]
@hsyed : I assume you mean "the ijar step is currently just cp'ing so it *cannot* be merged in", right?

@cushon : unfortunately, I can't recall or find that discussion anywhere :( I think in Bazel, that is okay because `proto_library` doesn't have an implicit dependency on anything Java, so there won't be a dependency cycle, but the internal version does, so we need to do some tricks to make that work and the collateral damage would be that the internal version wouldn't support Kotlin. I think we could live with that.

**Comment by lberki** [comment]
Ugh, I was just made aware that the extra dependencies required by protocol buffers make the above solution impractical :( I can think of a few options:

* Write an "ijar-for-Kotlin" binary that strips out the Kotlin-specific parts from jars as a step to be run before or after ijar. We'd need to tweak the Skylark API for Java a bit, but I think that's doable? (paging @iirina for confirmation)
* Check in the source files generated by the proto compiler, thus cutting the dependency that way.

I'd prefer the former one, because then people who don't use Kotlin don't pay the price to support it.

Any more clever ideas?

**Comment by hsyed** [comment]
@lberki I guess i got the convention mixed up. I’ll update the pr.

Do we need to unpack the proto for certain corner cases ? @cushon what did you find inside d1? d2 is a string array and does have a different shape for inline methods.

**Comment by hsyed** [comment]
Aah not a return type, `donkey` returns `Int`.

**Comment by hsyed** [comment]
@lberki an ijar-for-Kotlin would have to share most of the code of the existing tool no ?  Did you have an alternate approach in mind .  

> Write an "ijar-for-Kotlin" binary that strips out the Kotlin-specific parts

 don't you mean "retains some Kotlin-specific parts"  ?

**Comment by lberki** [comment]
I meant an ijar-like tool that runs *in addition* to ijar. Reading again, however, it doesn't make much sense: AFAIU the problem is that ijar strips out too much and not too little.

**Comment by hsyed** [comment]
@cushon do you see anything essential in d1 ?

 I can try and get conformation from Jetbrains on the approach for inferencing off the d2 annotation field.

**Comment by cushon** [comment]
> I meant an ijar-like tool that runs in addition to ijar. Reading again, however, it doesn't make much sense: AFAIU the problem is that ijar strips out too much and not too little.

@lberki - what about a separate ijar entry point that included kotlin support, which could be used instead of the regular ijar if kotlin support was needed?

>  do you see anything essential in d1 ?

@hsyed - my understanding is that d1 contains the actual metadata as a serialized proto, and d2 is arbitrary constants that are referenced by the metadata? The comments in [1] say:

<blockquote>
<dl>
<dt>d1
<dd>Metadata in a custom format. The format may be different (or even absent) for different kinds.
<dt>d2
<dd>An addition to [d1]: array of strings which occur in metadata, written in plain text so that strings already present in the constant pool are reused. These strings may be then indexed in the metadata by an integer index in this array.
</dl>
</blockquote>

But if you can get clarification from jetbrains on the best approach here, and whether the metadata is something that it makes sense to depend on in other tools (rather than being an unsupported implementation detail) that would be welcome.

[1] https://github.com/JetBrains/kotlin/blob/master/core/runtime.jvm/src/kotlin/Metadata.kt

**Comment by hsyed** [comment]
I've created an issue on the [Kotlin Youtrack](https://youtrack.jetbrains.com/issue/KT-22938) for feedback.

**Comment by hsyed** [comment]
I glossed over the “d2 is a symbol table” part :( 

@cushon in the multiple entry point approach were you thinking of adding a visitor extension (hiding proto behind the visitor) ? If so, what about dynamic loading an extension instead via flags ?

**Comment by hsyed** [comment]
@lberki @cushon There is an update on the [Kotlin Youtrack ](https://youtrack.jetbrains.com/issue/KT-22938). 

It seems we would need to unpack the proto for now. For the future I proposed a new annotation for ABI elements so we can figure out what to keep in a context-free way. Please give some feedback to JB on this one if you think we can make use of this.

I'm really hoping we can update the ijar tool for inlining in the short term, if we can get this in the Bazel Kotlin UX will rival the native Java rules.

**Comment by hsyed** [comment]
FYI buck use a [JVM based ABI tool](https://github.com/facebook/buck/tree/6043596e11bee8dc5119a78c89a8c744bd9e6868/src/com/facebook/buck/jvm/java/abi).

**Comment by hsyed** [comment]
@cushon @lberki It would be a suitable short term compromise for rules_kotlin to require the user to override the `java_toolchain` providing a modified version of the ijar tool. 

We could then create a custom build of the ijar tool in rules_kotlin or here. 

`java_toolchain` needs to be overridden via a define, slightly ugly. This could could be made seamless if the native ijar tool was picked up as a skylark toolchain. This could then be made part of the rules_kotlin workspace boilerplate. 

So, I propose

1) Make the global ijar tool discoverable in bazel core via a skylark toolchain type specifically for the ijar tool. 
2) Create a custom version of the ijar tool in rules_kotlin (or preferably here).

It would take me quite some time to skill up to make the changes to the ijar tool. If for 2 we think the tool should being in `rules_kotlin` I can do the preliminary work of setting up the build and including the proto srcs.

**Comment by lberki** [comment]
Erm, sorry for the radio silence. I don't think making `ijar` a toolchain is a good idea, because then how would Bazel know to use the regular ijar for Java rules and the Kotlin ijar for Kotlin rules?

I'm wondering if it's possible to use `java_common.create_provider` as the way to add Kotlin jars into the Java providers. That way, you can set `use_ijar` to false, then use a Kotlin-specific ijar tool to produce the compile-time jars. How that's implemented is up for discussion -- we can add a flag or two to ijar to make it do only the things that don't affect Kotlin and then a second tool would do the Kotlin-specific processing. Is that feasible?

**Comment by cushon** [comment]
> Erm, sorry for the radio silence. I don't think making ijar a toolchain is a good idea, because then how would Bazel know to use the regular ijar for Java rules and the Kotlin ijar for Kotlin rules?

I think the idea was not to have a separate ijar that only works for kotlin, but to have a kotlin-aware ijar that works with both kotlin and non-kotlin rules, and has a dependency on proto. If someone didn't want their host tools to depend on proto the wouldn't be able to use kotlin.

But your `java_common.create_provider` sounds workable to me.

> we can add a flag or two to ijar to make it do only the things that don't affect Kotlin and then a second tool would do the Kotlin-specific processing.

I think the second tool is the important part. `ijar` is going to remove all code attributes, and kotlin needs the implementation of inline methods to be preserved. So if we remove the parts of ijar that break kotlin I don't think there's anything left :)

**Comment by cushon** [comment]
> Indirecting via a `kt_jvm_import` is not the way forward, this indirection makes integration with other java common infrastructure more complicated (think android rules as well)

@hsyed can you elaborate? `java_common.create_provider` provides first-class access to the provider that `java_import` uses. Other rules (like the native Android ones) shouldn't care whether a dependency is a `java_import` or a skylark rule using `java_common.create_provider`. If that isn't true, it's something Bazel team is probably interested in improving.

**Comment by hsyed** [comment]
@cushon `java_common.create_provider()`  is fine, I didn't mean to suggest that it needs to be updated (Infact the last 2 releases have added things that I was looking for (neverlink and I think exports , I was emulating exports).

I wanted to highlight the naked jar scenario (without providers attached). Hava a look at the sheer amount of [ceremony and bloat](https://github.com/johnynek/bazel-deps/blob/master/dependencies.yaml) introduced because of scala to the dependency format of the  bazel-deps tool.

**Comment by lberki** [comment]
And we want to support people using a Kotlin jar without knowing that it's a Kotlin one? Because it we can expect that knowledge, `kotlin_import` is the right thing to do because it keeps Kotlin-specific knowledge in Kotlin rules, where it belongs.

**Comment by hsyed** [comment]
> And we want to support people using a Kotlin jar without knowing that it's a Kotlin one?

Yes and no. The no element is that we should attach providers to all jars. For a large repo it is generally best practise that deps should be indirected via `//third_party` this should be the case for Kotlin as well. Beyond best practise Kotlin has 4  facets that have to be modelled in the rules `common`, `jvm`, `native` and `js`. The later 3 are implementations of `common`.  All of the 4 facets can be encoded into jars so for everything but the `jvm` we _need_ a provider.

The yes element is specifically about the `jvm` rules being able to use both `JavaInfo` and `KtInfo` as deps. We don't "need" to model `KtInfo` for external deps. The rules currently do because they have to to force select `full_compile_jars`. The Java rules can consume Kotlin jars without a `KtInfo` and vice versa so why complicate things ? The java rules don't need to know they are dealing with Kotlin jars and the Kotlin `jvm` rules don't need to know they are dealing with java jars. Only that they have a `JavaInfo`. 

One of design requirements for Jetbrains was to use Kotlin to in-place convert  Intellij platform code to Kotlin with a code converter -- at a class level (not a package level) without breaking any of the Java code. So not making the distinction in the providers is approved by design :). 

The only spanner in the works is that two simple elements -- the bytecode body of an inline method and a metadata file need to be retained by the ijar tool. Tackling the JVM element so that we have the same level of support as native java support is what I am focusing on at the moment.

**Comment by hsyed** [comment]
I took a look at updating bazel-deps as we could use it now to manage kotlin artifacts -- I have it partially working but I time boxed the task and due to the amount of further work required it is worth it. The time would just be spent adding layers of indirection needed to turn off ijification of kotlin artifacts.

**Comment by johnynek** [comment]
@hsyed do you want to make a WIP pull request to bazel-deps to share what you have now? I’m happy to help you get this in a mergable state.

**Comment by hsyed** [comment]
@tomlu @cushon  I'm responding to the comments with  regards to the discussion that occurred in [this issue](https://github.com/bazelbuild/bazel/issues/4846).  

In response to writing out "Target Label" in the manifest I will add that, however there is then the `kt_jvm_import` problem. External jars won't have the Target Label annotation, I'd have to visit and repackage the Jar which means starting up a Kotlin Worker unecessarily.

The Kotlin ijar use case is straight forward on the JVM and has no comparisons to the complexity that Scala macros bring.  I'll summarise again here. Kotlin's advanced features are implemented on top of `CLASS` retention annotations and placeholder bytecode is used for inline methods. All the ijar tool has to do is leave these in the `hjar`. These can be discovered via context free visitation (well the metadata annotation needs to be unpacked with proto). 

I wish I could get the time to update the tool myself but I am swamped on many fronts -- I am maintaining the intellij bazel Kotlin plugin aspect, the rules themselves and I am doing most of this in my free time. I wish I had the time to make changes to the Ijar tool but it's too much up skilling up for me. I can understand the flow of the ijar tool roughly but I can't make the required changes since I am not familiar with the ins and outs of the bytecode spec. Once the ijar tool is updated with the necessary visitation logic I should be able to help maintain it going forward.

What can I do to get the ball rolling on this ? The issues discussed above with regard to cyclic dependencies etc could be sidestepped if we copybara the ijar tool into the kotlin repo and implement a variant of it there, we could then bring it back at a later date ?

**Comment by tomlu** [comment]
> I'd have to visit and repackage the Jar which means starting up a Kotlin Worker unecessarily

You can use this in your import rule to stamp the jar without using full ijar: https://docs.bazel.build/versions/master/skylark/lib/java_common.html#stamp_jar

**Comment by hzsweers** [comment]
FYI - there's an official metadata parser library now: https://github.com/JetBrains/kotlin/blob/master/libraries/kotlinx-metadata/jvm/ReadMe.md

You can also use the metadata annotations to understand whether or not certain methods are `inline`. In [Moshi](https://github.com/square/moshi)'s recent kotlin code gen support, we use a helper library called [kotlin-metadata](https://github.com/Takhion/kotlin-metadata). This library is what [ROOM](https://developer.android.com/topic/libraries/architecture/room) uses for its kotlin support as well to read things like parameter annotations.

Sidenote - parameter annotations haven't come up in the thread yet, but just want to note that those would be part of the ABI of a kotlin class. Same with internal vs public visibility changes (both `public` in the JVM).

**Comment by kevin1e100** [comment]
Thanks for these pointers.  While the metadata reader is handy, we're facing the dilemma that ijar is written in C++ and hence can't use it directly.  We've been actively thinking about what to do here but none of the solutions we've come up with are particularly appealing.

I'll note that parameter annotations (of non-private methods) shouldn't be stipped by ijar, and public methods should remain public.  If you're observing issues around parameter annotations or internal visibility, could you please provide an example?

**Comment by kevin1e100** [comment]
Quick update here: commit 088d8de6584864adaec82712c8ace601404afd0b makes it so `META-INF/<module name>.kotlin_module` are no longer stripped, which should help with extension methods when using ijar, but `inline` methods are still unresolved.

**Comment by hsyed** [comment]
@kevin1e100 great news. 

A bit more detail on the mechanics of what needs to occur with the bytecode:

1. For inline methods we also need to keep the bodies or at least the signatures of `internal` elements that are used by the inline methods these should be annotated with `@PublishedApi`.
2.  `typealias`'s also breaks after ijarification. I haven't looked into why.
3. The metadata annotations will have to be rewritten during ijarification. I completely glossed over this one when I made the proposal. In general any elements in the annotations that aren't part of the ABI or could break would need to be removed.  Specifically the constant pool needs to be pruned so that all the strings for the stripped elements .are cleared out --e.g., when public method `foo` is dropped for a class the string `foo` should be removed from the pool. This also means rewriting the indexes.

I think the ijar tool is the right place to do this it's native and Kotlin worker isn't needed. It is a complex task though going through the annotation spec and figuring out what to strip, and also keeping in sync with the spec. The Kotlin compiler has the infrastructure to decompile Kotlin class files from ASM -> PSI Model. 

I tried my hand at stripping via ASM in the Kotlin builder worker. Applying basic stripping transformations was very fast. Buck uses the JVM (ASM) to [ijarify](https://github.com/facebook/buck/blob/e90c76e61995615fe75d310e639f4cac090c44cc/src/com/facebook/buck/jvm/java/CalculateClassAbiStep.java) so it could work as a global replacement ? 

In general I think doing this in the JVM is probably where we can get the most bang for our buck, JB can hopefully contribute the stripping heuristics needed in the compiler jar -- if they do the heuristics will be tested and adapt with changes to the spec. 

So the question is could we replace the global ijar tool with something ASM based  ? If JB contributes the heuristics in the compiler jar it does mean bazel core would have a dep on Kotlin. 

Adding the Kotlin dep to bazel core is probably the right strategy in the long run anyway imo. I mean this just for the JVM aspect. Reasoning: 

1. The Kotlin builder is a replica of the Java builder and mostly just to add one step to the compile phase (turn `.kt` files into `.class` files and kapt generation) -- all the other phases are duplicated effort.
2. We have to have a second set of worker processes running for no good reason consuming significant amounts of memory during builds. 
3. The Kotlin builder has to use javac for kapt and mixed mode compiles -- it would be better to make use of turbine. 
4. Compiler plugins for Kotlin need to be written for dep analysis -- unifying this with the existing infrastructure makes sense. At the moment I am using the JDK jdep tool to analyse the jar and parse the text report to build the dep artifact.
5. Google is implementing the Kotlin rules internally. This is such a waste of resource.... If we centralise the JVM aspect we can at least share the hard bits of the problem.

**Comment by kevin1e100** [comment]
Thanks for the additional details.  As mentioned in my previous comment we are thinking about how to fully support Kotlin in ijar, but the technical constraints are such that there are no easy solutions here.  I'm sorry I don't have a better answer at this point.  (As you point out, re-implementing ijar in Java appears to be one possible approach.)

I'll point out thought that ijar is meant to be a build performance optimization.  Thus, ijar support shouldn't be necessary to compile Kotlin with Bazel (but using ijar might speed things up).  Does that match your understanding, or is lacking ijar support hard-blocking something?  (Apart from java_import not working quite right for Kotlin Jars, which I was under the impression you were able to work around.)

**Comment by hsyed** [comment]
> Thus, ijar support shouldn't be necessary to compile Kotlin with Bazel (but using ijar might speed things up). Does that match your understanding, or is lacking ijar support hard-blocking something

Having to `kt_jvm_import` was a minor inconvenience. We are gradually moving towards structuring our JVM code like go packages and we are porting these to Kotlin by the day. So not having ijarification gets incrementally more painful by the day -- some of our devs have moved from MacBooks to workstation hardware. So it's mostly about the dev UX now.

I'm going to look into adding incremental compilation support to the Kotlin builder when I get some time as this might be enough to get the UX to a point where the lack of ijarifaction won't be as painful. This is not officially supported by the command line compiler so I'd be wrapping the internals.

**Comment by hsyed** [comment]
Jetbrains is going to add support to the compiler to produce ABI jars. [Here is the youtrack issue](https://youtrack.jetbrains.net/issue/KT-25128) -- Please chip in case I missed something.

**Comment by hsyed** [comment]
The design of abi jar could produce the ABI jar and full jar in parallel or I guess in stages. 

I have not yet used the worker protocol for  multiple actions to the kotlin builder. Would it be possible to return the two artifacts as they become available ?

**Comment by cushon** [comment]
> I have not yet used the worker protocol for multiple actions to the kotlin builder. Would it be possible to return the two artifacts as they become available ?

No, workers don't have any special support for that. The interface jar would need to be generated by a separate action to be able to start downstream actions as soon as it was generated and without waiting for the full jar.

**Comment by cushon** [comment]
> I can't find how to register a custom ABI jar or register the generating tool in a rule. I assumed java_common.create_provider would enable this somehow.

This is supported with [JavaInfo.compile_jars](https://docs.bazel.build/versions/master/skylark/lib/JavaInfo.html#compile_jars). Note that `java_common.create_provider` is deprecated.

**Comment by cgruber** [comment]
Pinging on this, as this is starting to become a complication for us, especially if it requires special configuration for maven deps that happen to be Kotlin.

**Comment by lberki** [comment]
/cc @kevin1e100  @cushon 

So, getting up to speed on this, let me see if I understand the state of the world correctly: apparently, Kotlin requires a `.kotlin_module` file that is a serialized protobuf of a not-documented-but-available-in-the-code format: 

https://github.com/JetBrains/kotlin/blob/ee50aec7342ab375a105bb318e6501135e859028/core/deserialization/src/descriptors.proto

and ijar currently cannot depend on protocol buffers for Complicated Reasons. So I those Complicated Reasons (I believe related to bootstrapping) are fixed, can we munge the protos in `.kotlin_module` appropriately?

**Comment by cgruber** [comment]
I had thought you could just interrogate the bytecode for the @Metadata annotation, which contains in it the cue (at least from my quick looking at the bytecode) that this is an inline function.  We could probably avoid all sorts of proto munging if we just pivot off that, as I think @cushon mentioned above.

**Comment by cushon** [comment]
One option is to use a kotlin compiler plugin to add a magic `@com.google.devtools.ijar.KeepForCompile` annotation to the inline methods that you want to keep, see: 91b86cc58930148e0d968c3c393f94faf580545b.

@cgruber isn't the information we need from the `@Metadata` annotation encoded as a string of serialized protobuf data?

**Comment by cgruber** [comment]
Ah, I'm not sure. I was looking through and it was binary data, but I wasn't sure what form.  I was just verifying that a signal existed in the .class file.   Maybe you do need pb parsing. :/  Though you could probably hand-scan for it, however much that would suck, if you were super avoiding the protobuf dep.  Ick.

**Comment by cgruber** [comment]
As to using `@KeepForCompile` our problem right now is stdlib jars like kotlin-stdlib and kotlin-test which have reified inline types, but we have no control over the source.  If you want to get jetbrains to do that, it'd be neato. :D

**Comment by kevin1e100** [comment]
To be clear ijar at this point isn't able to process Kotlin Jars correctly,
including (in particular) the Kotlin standard library.  JetBrains is
working on a way to generate ABI Jars with kotlinc (see
https://youtrack.jetbrains.net/issue/KT-25128 mentioned earlier in this
issue), which can be used instead.  While this is promising,
https://youtrack.jetbrains.net/issue/KT-29471 suggests there's more work to
do here.

**Comment by davidsantiago** [comment]
Is it crazy to reverse the jar-stripping responsibility? Right now, my understanding is that the java_* rules strip jars that get given to them, and that causes a problem when some jars don't get stripped nicely. Would it make any sense to make the java_* rules instead offer a stripped jar as an implicit output, and then be more conservative in their inputs, using the ABI-jar if it's available and if not just using the full jar and taking the performance hit? Possibly configurable. 

Seems to me like this would make things more forgiving in a general way for other languages, so that knowledge of the language a jar was created in would not need to be transmitted all up and down the stack (for example, all of the Maven-based transitive dependency tools currently need to be told by the user to treat a dependency as having come from another specific language).

My expectation is that there is some subtle or not-so-subtle reason this couldn't work, but I thought I'd float it, because I've been dreaming of this so much lately.

**Comment by cgruber** [comment]
@kevin1e100  - that much is clear.  That's sort of the point.  Getting kotlinc to output abi jars is helpful to a point, but unless those are also shipped with maven, and we download those (which is hard because of the subtleties of repository_ctx.download) using pre-packaged abi jars isn't terrible satisfactory.  (Unless I'm wrong and there's a way to do a repository rules donwload action that is optional, and doesn't fail the build if hte file isn't there.  Otherwise we end up doing weird "fetch_kotlin_abi" optional configuration and it all gets messy, which is frustrating when the metadata is right there in the jar). 

I'd be ok turning off ijar on kotlin jars, but adding that sort of feature to maven integration, for instance, is a huge hassle without using kt_jvm_import, which then creates a dependency on the rules_kotlin, when what we really just need is a signal to not ijar.  If java_import had a (noijar = True) option, we could just use that and be done.

Is your comment just a clarification, or are you expecting that ijar will never support kotlin jars?  That would require some rethinking on all our parts about how to stage the tooling.

**Comment by ZacSweers** [comment]
ABI jar generation is available (though maybe incomplete?) Since 1.3.20. it
just runs as an alternative mode in the compiler similar to kapt stubs.
Reading the metadata protobuf is possible and technically a public API now,
albeit rough around the edges. That may be more appropriate if you want
super tight integration, but the public API is an unstable project called
kotlinx-metadata. It's also technically only JVM, and I'm not sure what
language ijar in Bazel wants to speak.

**Comment by lberki** [comment]
@cushon : does the "ask compiler to add an annotation" approach work with imported jars?

**Comment by cushon** [comment]
@lberki I think `.kotlin_module`s are unrelated to this bug, ijar already doesn't delete those entries from jars: https://github.com/bazelbuild/bazel/blob/f21ee4d114d31787bc9eda6ef90cad93d6b7fa85/third_party/ijar/ijar.cc#L106

IIUC the problem is that we would like it to not strip the implementation of `inline` methods, and the knowledge about which methods are `inline`s is encoded inside a serialized proto inside a `@Metadata` annotation in the code.

The only supported way to read that annotation is using the JVM kotlinx-metadata library, which we can't use because ijar is implemented in c++. And we can't access the proto directly because of the bootstrapping issue with c++ protos.

**Comment by lberki** [comment]
Grump. I think the protobuf issue can be solved with an appropriate amount of yak shaving, question is, do we do that, or do we use kotlinx-metadata? I can imagine an approach where imported jars need to be tagged as containing Kotlin code and then we'd dig out the metadata using kotlinx-metadata then tell ijar to do things differently.

It would involve more moving parts than just directly looking at the proto, though, and having to tag Kotlin jars as Kotlin jars isn't an especially nice user interface.

**Comment by cgruber** [comment]
Personally, I'd vastly prefer to not use kotlinx-metadata.  Staging the tooling in such a way as to have the kotlin infrastructure in place, without having it contaminate intermediary tooling, ends up being really crappy.  I'm having to jump through a fair number of hoops to keep the bazel_maven_repository stuff from having a hard dep on the kotlin rules (which don't yet support 1.3 anyway), as there's all sorts of extra workspace complexity to set up (or you end up with multiple definitions of the kotlin rules, because you import them, but the bazel/maven integration also imports them, and it gets weird.

So my preference is definitely to have as dependency-free an approach as possible, which is why just reading the @Metadata and interpreting it in the ijar tool was so attractive.  That said, I didn't understand the proto boostrapping problem. Also, I think we should push on Jetbrians to firm up the contract around the metadata proto.  If the current answer is "kotlinx-metadata" then I think that makes a really wierd situation for tooling and I think Jetbrains could be amenable to having a more data-format driven set of guarantees.

**Comment by cushon** [comment]
> I can imagine an approach where imported jars need to be tagged as containing Kotlin code and then we'd dig out the metadata using kotlinx-metadata then tell ijar to do things differently.

That would definitely work. The main drawback is that it requires knowing which jars contain kotlin, which is arguably not ergonomic. If we know which jars contain kotlin we could also just not run ijar on them...

**Comment by cgruber** [comment]
So, it'd be nice if we could get the benefits of ijar - turning it off for kotlin jars is one thing, but we're going to start feeling that pain as more and more of our codebase is turned over to kotlin, and the more we rely on maven-published libraries we have made, but decided to make github the source of truth.  It can work, but it would be sad. 

Finding out what jars contain Kotlin could be a matter of just if we encounter an @Metadata annotation, we're in Kotlin land, so alter accordingly.  That doesn't track with "don't run ijar" but does track with "do ijar differently".  I'm presuming ijar ends up scanning at least linearly, so having a bail-out path when it encounters that (or an alternate treatment) is an option. I don't know what sort of penalty we would be paying by not avoding running ijar in the first place, vs. the mid-ijar bailout path.  

For the maven integrations, I'm already exploring ways of just attributing "this is kotlin" in the dep declaration, and swapping the logic. I would highly rather NOT do that in the long-run, just to tweak bazel's use of ijar, hence my advocacy for ijar to do the right thing here.  But that's the workaround - we use a different kt_jvm_import which attaches a JavaInfo to that target that just says no to ijar. 

But that introduces complexities or hard dependencies on the kotlin rules, which is a big part of what we're trying to avoid here.

**Comment by ZacSweers** [comment]
> Finding out what jars contain Kotlin could be a matter of just if we
encounter an @metadata <https://github.com/metadata> annotation

Actually there's a higher level option - there will be a .kotlin_metadata
file present in the jar resources with module information. You could use
the presence of that alone as a sentinel if you wanted

**Comment by kevin1e100** [comment]
To an earlier point, Bazel's java_* rules do distinguish compile-time Jars,
and other rules can choose to distinguish them from the "real" Jars as
well.  Because of the limitations with ijar discussed here it's just not
possible at the moment for Kotlin build rules to use ijar in particular to
generate these compile-time Jars.  Do keep in mind though that the concept
of separate compile-time Jars is largely a build performance optimization.

With the compiler plugin that was mentioned a few times it'll be possible
to directly generate these compile-time Jars instead, without ijar.  Indeed
that won't work for prebuilt Jars, but one can use the full prebuilt Jars
as compile-time Jars in that case.  Bazel's java_import rule can't be used
for that, but it's possible to write a roughly equivalent rule in Starlark,
which was done for Kotlin to my knowledge.  What benefits are you expecting
from being able to ijar prebuilt Jars?  Is it a big advantage to be able to
use java_import instead of the Kotlin-specific rule?

**Comment by cgruber** [comment]
Yeah, that's how the kotlin rules do it now.

One key question here, that is maybe valuable to clarify solutions, is: are there performance gains from using abbreviated ijars later in the compilation steps?  That is, are there I/O gains to be had?  Or is this way more about cache hits on frequently changing stuff, so it really only makes so much sense on the source side, in which case tying kotlin source to the kotlin compiler is already a thing, so borrowing their ABI production in place of ijar is viable.  If there's little gain from ijar-ification of maven-supplied jars of any kind, then we could just drop this, not use java_import, and just make an import that passes in the real jar as the compile-jar.  That could work. 

The benefit of using java_import is really only in case there are sufficient to using ijar on pre-built jars.  If so, then the benefit that it lets tooling that is agnostic to Kotlin (say, the bazel/maven integration) pass along kotlin jars and have bazel do the right things, as opposed to having to swizzle in kotlin-specific logic, and possibly additional configuration.  But I'm starting to question my fundamental assumptions. 

I had assumed that the fact that java_import does ijar implied that there was a benefit to doing so. Are the benefits maybe less than that implication might suggest?

**Comment by cgruber** [comment]
Part of your question I missed answering though - the advantage of using java_import (or a non-ijaring but universally applicable alternative) to using anything form the kotlin_rules is that it creates inter-workspace dependencies that are really awkward (the effect of swizzling in the kotlin import into the bazel stuff).  Cross-workspace interaction, version management, etc., is tricky and problematic at the present time, so I'm going to fairly great lengths to avoid them, so end-consumers of the bazel maven integration workspace don't have to do extra crap that's just to align workspace assumptions, etc.

**Comment by davidsantiago** [comment]
@kevin1e100 That sounds very encouraging, and something like what I was hoping for above. In this Starlark `java_import_precompiled_jar` rule, let's call it... would that be possible to write it so that it interoperates with the existing Java rules? If so, that would let Maven dependency managers just output targets that use that rule and not have to worry about splicing in kotlin/scala/etc stuff, as cgruber just described.

**Comment by kevin1e100** [comment]
It should be pretty straightforward to write a Kotlin-agnostic Starlark
rule that behaves similar to java_import but skips ijar, yes.  The rule
largely just needs to return a JavaInfo provider with the imported Jar as
both the output_jar and the compile_jar.  I'm not sure about the build
performance impact of that compared to using ijar, but for (assumably
infrequently changing) prebuilt Jars I'd hope it's fine.

I still agree with the sentiment that it would be nice if java_import would
"just work" for Jars compiled from Kotlin etc., and I think it's worth
keeping this bug open to still try and address that somehow.  A possible
stopgap I thought of in the meanwhile was to make ijar a no-op for any
class with a Kotlin @Metadata annotation.  That obviously defeats the
purpose somewhat, but it's "safe" and would make java_import work with
Kotlin Jars (barring other issues).

**Comment by cgruber** [comment]
Yeah, I think the safe-java_import would be slightly preferable, if possible, since it wouldn't disable ijar for non-kotlin jars.  In the immediate term, I guess I'll make the maven integration use an ijar-disabled import rule until "safe" ijar ships.  I think the real question is, how much gain do we get from ijar for infrequently changing builds.  If it turns out "not much" then maybe this gets deprioritized.  If so, then it's probably still worth pushing on this long-term.

Not sure how much the bazel team is set up to do performance testing.  Our setup is kotlin-heavy, so it'd not be terirbly useful for comparison testing between ijar/non-ijar, since the ijar case currently borks.  But we could probably gin up something for testing purposes.  Just concerned if it's too trivial we won't see meaningful numbers because the scale isn't sufficient to see them, rather than because there is or isn't gains at-scale.

**Comment by cgruber** [comment]
It just occurred to me that we might end up with an issue with aar_import.  Does native.aar_import do ijar against the internal classes package?  That might need a more robust workaround (unless the android_rules port to starlark is imminent)

**Comment by kevin1e100** [comment]
Worth checking.

Just for the record, one other issue with importing prebuilt Kotlin Jars
can be that they typically depend on kotlin-stdlib.jar and friends.  That's
completely orthogonal to whether ijar is used or not, but it can be a
gotcha.

**Comment by cgruber** [comment]
Well, in the case of my work, I'm making maven integration, which automatically parses pom information to create the correct bazel deps information for inter-artifact stuff.  So using it would run kotlin-stdlib.jar through the same system.  I don't think that'll bite us unless someone brings it in via another path (always a problem with any management of external jars in bazel, but not a particularly worse problem than usual)

Most of the commenters here are looking at dependency-aware integrations, but the rules_kotlin fellow wasn't... but rules_kotlin may assume that it's an orthogonal problem handled by another set of rules, or by the main workspace owner.

**Comment by cpsauer** [comment]
Hey, all! Figured I'd toss up a quick PR that disables ijar in java_import by default. The goal is to unblock its usage for Kotlin JARs enable reunification around java_import if you folks  wanted it. 

So please see #14967. Would love it if you'd weigh in, lmk what you think, or add to it! 

Backstory:
I'd been fixing some other things in java_import around Android/ProGuard and ran across this issue when I noticed that my fix wasn't automatically applying to rules_jvm_external. (See #14966 and https://github.com/bazelbuild/rules_jvm_external/issues/672.) From the volume of discussion and all the java_import re-implementations, it seems like there's a strong need here, so, like you, I'm hoping Bazel folks will up-priorize a real fix to ijar! But in the meantime, it seemed like disabling ijar with a togglable attribute was the right stopgap measure, as you all had proposed above. I'd just added a (private) attribute to java_import for my other PR, so I figured I might as well take a quick crack at this one. I'm not (currently) a Kotlin user or anything, but if felt like the right thing to do. So I could use your help making sure it works for you and integrating when the time comes!

**Comment by cpsauer** [comment]
For (future) readers of this issue: Plenty of additional discussion over in #14967.

To answer a dangling question above: It doesn't like they bothered with ijar in aar_import.
Presumably that's for the same reasons as above, and that you all didn't, namely the build-caching optimization it conveys isn't that relevant to prebuilt code.

**Comment by cushon** [comment]
To recap some of the reason ijar exists for `java_import` from (https://github.com/bazelbuild/bazel/pull/14967#issuecomment-1062505959, https://github.com/bazelbuild/bazel/pull/14967#issuecomment-1072561368), it makes incremental builds faster when the jar changes but the ABI doesn't, and it makes the jars on the compile-time classpath much smaller. Prebuilt jars may not change very often, but `java_import` is also used with generated jars (e.g. together with `genrules`). Those jars can very large and change often. Disabling ijar by default for `java_import` doesn't look like a viable approach to me.

**Comment by cushon** [comment]
To expand on the previous comment, the option that @kevin1e100 brought up in https://github.com/bazelbuild/bazel/pull/14967#issuecomment-1062473166 of having ijar not strip classes annotated with `@kotlin.Metadata` seems like the most promising option that's been discussed to me.

We might also consider adding more validation to `java_import`, or example to detect mistakes like forgetting to add the kotlin standard library dep when importing kotlin jars, or detecting attempts to use `java_import` on jars that contain kotlin inline functions (until this bug is fixed).

**Comment by cpsauer** [comment]
@cushon and @kevin1e100, a question coming from a place of true ignorance about the state of a different alternative: 

I'd read @kevin1e100's comment above (https://github.com/bazelbuild/bazel/issues/4549#issuecomment-462509693) about generating ABI jars (equivalent to ijars?) via an official kotlinc plugin. I'd also seen that Kevin had contributed a fix to the other blocking issue he'd linked (awesome. Do we know the status of that, or if it is a feasible alternative? That is, would it also work for Java and therefore be an ijar alternative that would work with Kotlin? (I got as far as seeing that there was maybe some experimental ABI jar support in rules_kotlin, maybe based on it.)

**Comment by kevin1e100** [comment]
@cpsauer the jvm-abi-gen plugin I'm guessing you're asking about produces ijar-like artifacts _when compiling Kotlin sources_. I don't believe isn't usable with Jar files the Kotlin compiler previously produced, so it's not helpful in the context of `java_import` afaik.

**Comment by cpsauer** [comment]
Oh! Got it. Bummer Thanks, @kevin1e100, for taking the time to reply.

**Comment by cushon** [comment]
I had a quick look at kevin1e100's `@kotlin.Metadata` idea (https://github.com/bazelbuild/bazel/issues/4549#issuecomment-1081386387), and it seems feasible: https://bazel-review.googlesource.com/c/bazel/+/195202/1/third_party/ijar/classfile.cc

That will meant that ijar / `java_import` automatically don't strip Kotlin classes going forward, without needing any new attributes/configuration.

There's some possible future work around around more precise interface jars for Kotlin. To try to summarize, theoretically ijar could decode the proto in the `@Metadata` annotation and check for 'inline' modifiers:

* adding a protobuf dep to ijar might require some wrangling, I think we tried to add a protobuf dep to singlejar at one point and it caused problems, but I don't remember the details and it was a while ago
* ijar would need to preserve anonymous classes used in inline methods, I'm not sure if checking the `EnclosingMethod` attribute would be sufficient
* there was some speculation we might need internal heuristics

If someone wants to take a swing at that I think we'd want to try to accept that contribution, but it could be a bunch of work.

**Comment by kevin1e100** [comment]
That's amazing, thank you so much!

**Comment by cpsauer** [comment]
Wow! That's awesome. Thanks @cushon for getting it done.
(I'll update some of the other folks I'd seen blocked on it.)

**Comment by Bencodes** [comment]
Thanks for making that happen @cushon!
